[QgsQuick] Custom filepaths for external resource widget

sklencar opened the pull request:
* Added support for defining defaultPath:
 while taking pictures with external widget's photoPanel. Default path is taken from a widget config `path -> default path`. Also options `Relative to project path` or `Relative to default path` are taken into account. In addition, default path can also contain expression, which is evaluated in global, project and layer scope. Therefore these properties have been added while creating widgets.
* Loading PhotoCapturePanel only when needed:
The panel has been loaded with a form even if it wasn't needed at that moment. In addition, it was asking permissions for accessing camera on form load which was odd. Since the loader is used, its loaded after 'capture' action is initialised. 
* Long filenames wrapping
WrapAtWordBoundaryOrAnywhere is used now to prevent overflowing of a long name of files.

Commit:
Enabled using default path for external resources.
Value (filepath) can be relative to either project path or default root according field setup/config

Commit:
[QgsQuick] Added signal of event when a captured photo is confirmed. That enables custom handling of the action by externalResourceHandler.
Renamed Ok Button of PhotoPanel.

Commit:
[QgsQuick] Cleaning source of preview image after closing it
It was causing unpleasant effect when last taken photo has been shown for a short while before previewing another.

Commit:
[QgsQuick]
Evaluate and use defaultRootPath in a config of the external resource widget.
Added a function to evaluate expressions for QgsQuick

Commit:
[QgsQuick] Use Loader to load photo panel only when its going to be used.
Also prevents for asking camera permissions on form open, when photo panel was initialized as well.

Commit:
[QgsQuick] Better wrapping of a long filenames text in the external source widget if an image is not available.

Commit:
[QgsQuick] Use canonical path in function
Symlinks are used a lot on Android, therefore its better to use canonical paths.

Commit:
[QgsQuick] formatted code.

wonder-sk commented on src/quickgui/plugin/editor/qgsquickexternalresource.qml:
```diff
@@ -36,6 +36,30 @@ Item {
+      config["DefaultRoot"] ? config["DefaultRoot"] : homePath
+    }
+  }
+  property string prefixToRelativePath: relativeStorageMode === 2 ? targetDir : homePath
```
for absolute paths this should be probably empty

wonder-sk commented on src/quickgui/plugin/editor/qgsquickexternalresource.qml:
```diff
@@ -67,13 +91,13 @@ Item {
-    edge: Qt.RightEdge
-    imageButtonSize: fieldItem.iconSize
+  Loader {
+    id: photoCaptuePanelLoader
```
s/Captue/Capture/

wonder-sk commented on src/quickgui/plugin/editor/qgsquickexternalresource.qml:
```diff
@@ -97,7 +121,7 @@ Item {
       MouseArea {
         anchors.fill: parent
-        onClicked: externalResourceHandler.previewImage(homePath + "/" + image.currentValue)
+        onClicked: externalResourceHandler.previewImage(prefixToRelativePath + "/" + image.currentValue)
```
handling of absolute paths is missing (should be just image.currentValue). I would suggest to wrap it into a function that would return either "prefix/image.currentValue" or "image.currentValue" based on the absolute/relative path settings

wonder-sk commented on src/quickgui/plugin/qgsquickfeatureform.qml:
```diff
@@ -62,6 +62,17 @@ Item {
+        /**
+          * Called when clicked on the OK icon after taking a photo with the Photo panel.
+          * \param itemWidget editorWidget for modified field to send valueChanged signal.
+          * \param prefixToRelativePath Absolute path to the image extracted by current value.
```
This is not absolut path to the image right? just the path prefix?

wonder-sk commented on src/quickgui/plugin/qgsquickphotopanel.qml:
```diff
@@ -22,7 +22,12 @@ import QtGraphicalEffects 1.0
+  // Capture path
   property var targetDir
+  // Path to project
+  property var homePath
```
Is this needed? I can't see it used anywhere in the file

wonder-sk commented on src/quickgui/plugin/qgsquickphotopanel.qml:
```diff
@@ -22,7 +22,12 @@ import QtGraphicalEffects 1.0
   property var targetDir
+  // Path to project
+  property var homePath
+  // Along with lastPhotName creates an absolute path to a photo. Its either project path or defaultRoot.
```
s/Phot/Photo/

wonder-sk reviewed (commented):
Because the path handling is not completely trivial, I would suggest to add an overview like this somewhere in the external resource widget source file (it took me a bit of time to figure out how all the things are supposed to work)
```
Overview of path handling
-------------------------

Project home path: comes from QgsProject::homePath() - by default it points to the folder where
the .qgs/.qgz project file is stored, but can be changed manually by the user.

Default path: defined in the field's configuration. This is the path where newly captured images will be stored.
It has to be an absolute path. It can be defined as an expression (e.g. @project_home || '/photos') or
by plain path (e.g. /home/john/photos). If not defined, project home path is used.

In the field's configuration, there are three ways how path to pictures is stored in field values:
absolute paths, relative to default path and relative to project path. Below is an example of how
the final field values of paths are calculated.

  variable         |     value
-------------------+--------------------------------
project home path  |  /home/john
default path       |  /home/john/photos
image path         |  /home/john/photos/img0001.jpg


   storage type          |   calculation of field value    |      final field value
-------------------------+---------------------------------+--------------------------------
absolute path            |  image path                     |   /home/john/photos/img0001.jpg
relative to default path |  image path - default path      |   img0001.jpg
relative to project path |  image path - project home path |   photos/img0001.jpg
```

Commit:
[QgsQuick] External source widget - changes requested in review
Fixed test and added one test case for getting relative path.

Commit:
[QgsQuick] External widget update
Added function that returns absolute path of image according widget's config - takes into account relative/absolute setting for value and default path.
Added docs about how a path can be configured and used.

Commit:
[QgsQuick] External widget update
* removed lefover property
* corrected getAbsolutePath function

Commit:
[QgsQuick] Corrected function and fixed test
Do not use canonical path for non-existing file while getting relative path.
